### PR TITLE
jewel: build/ops: "osd marked itself down" will not recognised if host runs mon + osd on shutdown/reboot

### PR DESCRIPTION
**Pull request by shinobu-x**
http://tracker.ceph.com/issues/18906

### CODE REVIEW

**Comment by dachary** [comment]
Could you please update the commit so that it does not include anything but the addition of  **ceph-mon.target** at the end of the line. Also please make sure you mention

```
Conflicts:
	systemd/ceph-osd@.service: trivial context change
```
to clarify how the commit differs from the original. Thanks !

**Comment by smithfarm** [comment]
@shinobu-x This is the master commit: https://github.com/ceph/ceph/commit/7f4acf45dd0d86e7d9992a8c30e5876fb57b1914

This is the cherry-pick: https://github.com/ceph/ceph/pull/13492/commits/a4b1a22189a0b01de6e2c667a7602f7b983324f0

See how the cherry-pick differs? That's bad. The cherry-pick should do *only* what the master PR is doing, nothing more.

**Comment by shinobu-x** [comment]
@smithfarm Thank you for your advice. I will have a look more closely,

**Comment by shinobu-x** [comment]
@smithfarm I did not remember exactly what I did. But I'm clearly clear about that :) 

Usually I delete lines which are "<<<<<<<" and ">>>>>>>" and delete string which is "=======".

**Comment by shinobu-x** [comment]
@smithfarm Okey I'm now becoming better backporter hopefully :)

**Comment by shinobu-x** [comment]
What I had to is to:
DELETE

> Description=Ceph object storage daemon
> After=network-online.target local-fs.target time-sync.target

KEEP

> Description=Ceph object storage daemon osd.%i
> After=network-online.target local-fs.target time-sync.target ceph-mon.target

Mmm... Nothing difficult. But just only needs to open my eyes =)

**Comment by shinobu-x** [comment]
Cherry-picked only:

> After=network-online.target local-fs.target time-sync.target ceph-mon.target

**Comment by smithfarm** [comment]
> Usually I delete lines which are "<<<<<<<" and ">>>>>>>" and delete string which is "=======".

These lines are calling your attention to where the conflict lies :-) Git is not (yet) smart enough to actually resolve the conflict. That is up to you and your analytical skills.

**Comment by shinobu-x** [comment]
@smithfarm Yeah I should work git as well and a bit more smarter than him (maybe her) =)

$ git cherry-pick -x 7f4acf4
`error: could not apply 7f4acf4... systemd: Start OSDs after MONs`
$ git status
```
snip
#	both modified:      systemd/ceph-osd@.service
```
$ git diff c9ece04 d0a9f5c systemd/ceph-osd@.service
```
diff --git a/systemd/ceph-osd@.service b/systemd/ceph-osd@.service
index bd5a78b..0a43ee1 100644
--- a/systemd/ceph-osd@.service
+++ b/systemd/ceph-osd@.service
@@ -1,6 +1,6 @@
 [Unit]
-Description=Ceph object storage daemon
-After=network-online.target local-fs.target time-sync.target
+Description=Ceph object storage daemon osd.%i
+After=network-online.target local-fs.target time-sync.target ceph-mon.target
 Wants=network-online.target local-fs.target time-sync.target
 PartOf=ceph-osd.target
```

**Comment by shinobu-x** [comment]
@smithfarm Now I've got your point. Those lines are just attention. How conflicts need to be fixed is really up to myself =) Git has been innocent not :D

File: systemd/ceph-osd@.service [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [Unit]
-Description=Ceph object storage daemon
-After=network-online.target local-fs.target time-sync.target
+Description=Ceph object storage daemon osd.%i
```
Comment (smithfarm) [inline systemd/ceph-osd@.service]: This line is not changed in the master commit, so it should not be changed in the cherry-pick either.

Also, your commit message contains no description of the conflict or how you resolved it.

**Comment by shinobu-x** [comment]
@smithfarm Thanks so much for conveying information to me. That's also really really good one.
Please have a look here:
#1 Do ``cherry-pick``
Then I got conflict as expected :)
```
root@octopus:ceph [wip-18516-jewel]$ git cherry-pick -x 7f4acf4
error: could not apply 7f4acf4... systemd: Start OSDs after MONs
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'
hint: and commit the result with 'git commit'
```
#2 Do check what's conflicted
```
diff --cc systemd/ceph-osd@.service
index bd5a78b,0a43ee1..0000000
--- a/systemd/ceph-osd@.service
+++ b/systemd/ceph-osd@.service
@@@ -1,6 -1,6 +1,11 @@@
  [Unit]
++<<<<<<< HEAD
+Description=Ceph object storage daemon
+After=network-online.target local-fs.target time-sync.target
++=======
+ Description=Ceph object storage daemon osd.%i
+ After=network-online.target local-fs.target time-sync.target ceph-mon.target
++>>>>>>> 7f4acf4... systemd: Start OSDs after MONs
  Wants=network-online.target local-fs.target time-sync.target
  PartOf=ceph-osd.target
```
#3 Do modify conflict appropriately
```
diff --cc systemd/ceph-osd@.service
index bd5a78b,0a43ee1..0000000
--- a/systemd/ceph-osd@.service
+++ b/systemd/ceph-osd@.service
@@@ -1,6 -1,6 +1,6 @@@
  [Unit]
-Description=Ceph object storage daemon osd.%i
+Description=Ceph object storage daemon
- After=network-online.target local-fs.target time-sync.target
+ After=network-online.target local-fs.target time-sync.target ceph-mon.target
  Wants=network-online.target local-fs.target time-sync.target
  PartOf=ceph-osd.target
```
According to master (#13097), only declaration ``After`` was modified under ``[Unit]`` section.
Since that,  declaration ``Description`` doesn't have to be modified.

Modification ended up with:
 #a Keep existing line of ``Description``
  +Description=Ceph object storage daemon
 #b Modified ``After``
  +After=network-online.target local-fs.target time-sync.target ceph-mon.target

To make sure there is zero modification of ``Description``, here is difference between master and this branch:
```
root@octopus:ceph [wip-18516-jewel]$ diff systemd/ceph-osd\@.service ../../staging/ceph/systemd/ceph-osd\@.service 
2c2
< Description=Ceph object storage daemon
---
> Description=Ceph object storage daemon osd.%i
```
Since I DO not break Ceph because I do love it, please let me know if anything.
Thank you for your patience.

**Comment by smithfarm** [comment]
@shinobu-x Exactly right. Now, the question is how to describe this conflict in the commit message. Since this is a very trivial conflict and the resolution is obvious, one option is to simply write "trivial resolution" like this:

```
Conflicts:
	systemd/ceph-osd@.service (trivial resolution)
```

However, since you are learning this it might be a good idea to find out which commit introduced the conflict. You would find this out by doing `git blame systemd/ceph-osd@.service` in the master branch. Find the SHA1 of the commit that added `osd.%i` after "storage daemon". Then, in the commit message, you could write something like this:

```
Conflicts:
	systemd/ceph-osd@.service (jewel does not have $SHA1_OF_COMMIT_THAT_INTRODUCED_CONFLICT)
```

(of course, replace `$SHA1_OF_COMMIT_THAT_INTRODUCED_CONFLICT` with the actual commit hash that you found).

**Comment by smithfarm** [comment]
@shinobu-x By the way, once you know what you want to write, actually changing the commit message is very simple. Just checkout the wip branch corresponding to this PR, and then:

```
git commit --amend     # starts editor
git push -f origin $WIP_BRACH   # updates the PR
```

**Comment by shinobu-x** [comment]
@smithfarm Thank you so much. Please hold on. I will update you with exactly how I understand your kind message after quick sleep (sorry for this...). Thank you for your patience.

**Comment by shinobu-x** [comment]
@smithfarm Please check updated this PR. Because of your very kind and precise explanation, I've been getting more better understandings.

Nothing was changed but commit message of this PR, then pushed local into origin with a new message.

Thank you for letting me know what you are exactly thinking of.

**Comment by smithfarm** [comment]
@shinobu-x Thanks, that looks good. As I understand it, conflict resolution has three parts:

1. understand the differences (between master and the branch you are cherry-picking to) that cause the conflict (`git blame` is often useful here)
2. find the most minimal way to resolve the conflicts (or, sometimes, cherry-picking an additional commit or commits can eliminate the conflict)
3. in the commit message, describe the conflict and how you resolved it (this is to make it easier for other developers to understand what you did)

This was a good example for learning because the conflict was simple to understand and resolve. You could have done the backport without any conflict if you had cherry-picked 4179aa8 first, and then 7f4acf4 second. Often there is more than one way to proceed, and it's not always clear which way is the best way.

**Comment by shinobu-x** [comment]
@smithfarm Thank you very much for guiding me to right direction which was very exciting. I'm always super happy with ``do_some`` for ``Ceph``.

And now because of your great help I did it with more better way which was pretty good. I would be more flexible to do backport as you suggested in your previous message.

**Comment by smithfarm** [comment]
**This PR is targeting 10.2.8 so it must not be merged until the v10.2.7 tag has been added**

This PR passed an `upgrade/hammer-x` suite at http://tracker.ceph.com/issues/19538#note-7

**Comment by smithfarm** [comment]
Passed upgrade tests at [1] and [2]

[1] http://tracker.ceph.com/issues/19538#note-6
[2] http://tracker.ceph.com/issues/19538#note-7

**Pull request merged**